### PR TITLE
Various minor improvements to speed up by ~10%

### PR DESCRIPTION
**Pull request by kkmuffme**
Improve various minor things to speed up psalm performance by about 10%

No negative impact on memory consumption.

Will invalidate old caches (= most cache files created with any previous psalm versions will not be used anymore and new caches will be created on first run, since the cache hash changed for performance reasons)

### CODE REVIEW

File: src/Psalm/Internal/Clause.php [hunk 1]
```diff
@@ -106,11 +108,15 @@ public function __construct(
         } else {
             ksort($possibilities);
 
-            foreach ($possibilities as $i => $_) {
+            foreach ($possibilities as $i => $v) {
+                if (count($v) < 2) {
```
Comment (orklah) [inline src/Psalm/Internal/Clause.php:112]: Is this really useful? I'd expect sort to not execute anything if there's nothing to sort

Comment (kkmuffme) [reply src/Psalm/Internal/Clause.php:112]: I'd expect that to happen too, but alas, this is not what PHP does. Calling sort conditionally is much faster, according to the profiler.

This is something that could be improved in PHP core btw

File: src/Psalm/Internal/Scanner/DocblockParser.php [hunk 2]
```diff
@@ -87,15 +88,15 @@ public static function parse(string $docblock, int $offsetStart): ParsedDocblock
...
+            if (preg_match('/^[ \t]*\*?\s*@([\w\-\\\:]+)[\t ]*(.*)$/', $line, $matches, PREG_OFFSET_CAPTURE)) {
                 /** @var array<int, array{string, int}> $matches */
                 [, $type_info, $data_info] = $matches;
 
                 [$type] = $type_info;
                 [$data, $data_offset] = $data_info;
 
-                if (strpos($data, '*')) {
-                    $data = rtrim(preg_replace('/^[ \t]*\*\s*$/m', '', $data));
+                if (trim($data) == '*') {
```
Comment (orklah) [inline src/Psalm/Internal/Scanner/DocblockParser.php:98]: please use === here

File: src/Psalm/Internal/Clause.php (hunk 1 above)
Comment (orklah) [reply src/Psalm/Internal/Clause.php:112]: Ok :) Thanks then :)

Comment (kkmuffme) [reply src/Psalm/Internal/Clause.php:112]: https://github.com/php/php-src/issues/8887

File: src/Psalm/Internal/Scanner/DocblockParser.php (hunk 2 above)
Comment (kkmuffme) [reply src/Psalm/Internal/Scanner/DocblockParser.php]: There's some weird stuff going on in there, I just removed the commit for the docblock parser, as it's a time waste.

**Comment by kkmuffme** [comment]
Branch is ready to be merged.

Sorry for many force push, but the tests don't work for me locally. Some step by step guidance/link on how to run the tests locally (or where the composer is installed, even when on a remote server) would be appreciated.

**Comment by AndrolGenhald** [comment]
@kkmuffme I use `vendor/bin/paratest -f --max-batch-size=16` (tweak batch size to see what's fastest for your hardware), we probably should document that somewhere. Note that you may have to disable the callmap test due to #8162, I haven't had time to look into that yet.

**Comment by orklah** [comment]
Thanks!

**Pull request merged**